multi_partner_learning.py

```diff
@@ -0,0 +1,522 @@
+# -*- coding: utf-8 -*-
+"""
+Functions for model training and evaluation (single-partner and multi-partner cases)
+"""
+
+import os
+import pickle
+import keras
+from keras.backend.tensorflow_backend import clear_session
+from keras.callbacks import EarlyStopping
+import numpy as np
+import matplotlib.pyplot as plt
+import operator
+from loguru import logger
+
+import utils
+import constants
+
+
+class MultiPartnerLearning:
+
+    def __init__(self,
+                 partners_list,
+                 epoch_count,
+                 minibatch_count,
+                 x_val_global,
+                 y_val_global,
+                 x_test_global,
+                 y_test_global,
+                 multi_partner_learning_approach,
+                 aggregation_weighting="uniform",
+                 single_partner_test_mode="global",
+                 is_early_stopping=True,
+                 is_save_fig=False,
+                 save_folder="",
+                 ):
+
+        # Attributes related to partners
+        self.partners_list = partners_list
+        self.partners_count = len(partners_list)
+
+        # Attributes related to data
+        self.val_data = (x_val_global, y_val_global)
+        self.test_data = (x_test_global, y_test_global)
+
+        # Attributes related to the multi-partner learning approach
+        self.learning_approach = multi_partner_learning_approach
+        self.aggregation_weighting = aggregation_weighting
+        self.single_partner_test_mode = single_partner_test_mode
+
+        # Attributes related to iterating at different levels
+        self.epoch_count = epoch_count
+        self.epoch_index = 0
+        self.minibatch_count = minibatch_count
+        self.minibatch_index = 0
+        self.is_early_stopping = is_early_stopping
+
+        # Attributes for storing intermediate artefacts and results
+        self.minibatched_x_train = [None] * self.partners_count
+        self.minibatched_y_train = [None] * self.partners_count
+        self.aggregation_weights = []
+        self.models_weights_list = [None] * self.partners_count
+        self.local_score_list = [None] * self.partners_count
+        self.score_matrix_per_partner = np.zeros(shape=(self.epoch_count, self.minibatch_count, self.partners_count))
+        self.score_matrix_collective_models = np.zeros(shape=(self.epoch_count, self.minibatch_count + 1))
+        self.loss_collective_models = []
+        self.test_score = None
+        self.nb_epochs_done = int
```

Comment:
We should aim to simply the epochs attribute. We have three right now:

- `nb_epochs_done `
- `epoch_index `
- `epoch_count `

Maybe we should rename `epoch_count `to something like `max_epoch_count `and remove `nb_epochs_done `

Reply:
And `nb_epochs_done`= `epoch_index `?

Reply:
In multi-partner setting yes. In single-partner setting, this is derived from an attribute of the `EarlyStopping()` instance.